### PR TITLE
Hashable conformance and useful functions

### PR DESCRIPTION
**Pull request by letatas**
Added Hashable conformance, and reduce, map, filter and flatMap functions.

Tests included.

### CODE REVIEW

File: Sources/CountedSet.swift [hunk 1]
```diff
@@ -245,3 +259,50 @@ extension CountedSet: Sequence {
...
+        for (element, count) in backingDictionary {
+            if let (newElement, newCount) = try transform(element, count) {
+                result.update(with: newElement, count: newCount)
+            }
+        }
+        
+        return result
+    }
+
+    public func filter(_ inclusionCount: (Element, Int) throws -> Int) rethrows -> CountedSet<Element> {
```
Comment (kareman) [inline Sources/CountedSet.swift]: Shouldn't the closure parameter to filter return a Boolean?

**Comment by kareman** [comment]
Thank you, this could be very useful. But CountedSet is a Sequence so it already has the methods reduce, map, filter, and flatMap from the Standard Library. And they all (except for reduce) return arrays, not CountedSets. Dictionary and Set behave in the same way.

Users of this library will probably _expect_ map, filter and flatMap to return arrays because that's what similar Collections in the Standard Library do. But it is possible that most of the time they will _want_ these methods to return CountedSet. And with this implementation they can select which version of the methods to use by explicitly specifying the return type when calling them.

But it can lead to some confusing error messages:

```swift
let countedSet = CountedSet([1, 1, 2, 17])
let expected = CountedSet([0.5, 0.5, 0.5, 0.5, 1, 1, 8.5, 8.5])
        
let result1 = countedSet.map { (element: 0.5 * Double($0), count: 2 * $1) } // CountedSet<Double>
let result2 = countedSet.map { (0.5 * Double($0), 2 * $1) } // CountedSet<Double>
let result3 = countedSet.map { (0.5 * Double($0.0), 2.0) } // [(Double, Double)]
let result4 = countedSet.map { (0.5 * Double($0), 2.0) } // build error: Cannot invoke initializer for type 'Double' with an argument list of type '((Int, Int))'
let result5 = countedSet.map { 0.5 * Double($0.0) + 2 * $0.1 } // build error: No '+' candidates produce the expected contextual result type '(element: _, count: Int)'
let result6 = countedSet.map { 0.5 * Double($0.0) } // [Double]
```

The actual error for result5 is trying to add a Double and an Int.

It's hard to say what is the best way of providing this useful functionality without confusing the users with unexpected return types and strange build errors. Any thoughts, @0x7fffffff ?

**Comment by letatas** [comment]
Indeed, but I found that the default `Sequence` methods do not take advantage of the count of elements in the `CountedSet`. Maybe the solution would be to rename these methods: `countedSetMap`, `countedSetFlatMap`, `countedSetFilter` and `countedSetReduce`.

File: Sources/CountedSet.swift (hunk 1 above)
Comment (letatas) [inline Sources/CountedSet.swift]: The original `filter` function returns a boolean to include or not the element in the resulting collection.  As the `CountedSet` contains the number of element, the idea of the filter is to return the new count of the element, not only its presence. 

But as I commented, maybe we could just rename if `countedSetFilter`?

**Comment by kareman** [comment]
Yes, that should work. I've been trying to find out what the common practise is for this in Swift but apparently there isn't one.

To help these methods be more discoverable and play better with autocompletion, how about mapAsCountedSet or mapToCountedSet?

**Comment by letatas** [comment]
Good idea, the `*ToCountedSet` makes sense for `filter`, `map` and `flatMap`. And maybe `reduceCountedSet` for the last one?

**Comment by kareman** [comment]
Sounds good to me. 

About reduceCountedSet, it seems it does exactly the same as the built in reduce does now, do we really need it?

**Comment by kareman** [comment]
CountedSet is a sequence of (Element, Int) pairs, so the built in reduce has the signature

```swift
func reduce<Result>(_ initialResult: Result, _ nextPartialResult: (Result,  (Element, Int)) throws -> Result) rethrows -> Result
```

 If I comment out your implementation of reduce and change your unit test to this:

```swift
let countedSet = CountedSet([1, 1, 2, 3, 5, 8, 13])
let sum = countedSet.reduce(0) { $0 + $1.1 * $1.0 }
        
XCTAssertEqual(33, sum)
```

it still passes.

**Comment by letatas** [comment]
It seems I was mistaken ;) ! My bad...

**Comment by kareman** [comment]
Hey, no problem :) , and thanks again for the code. I'll leave this open for the moment in case @0x7fffffff has any input, if not I will merge it in later.

**Comment by letatas** [comment]
I just updated the PR according to what we said.

**Comment by 0x7fffffff** [comment]
@kareman Sorry for my absence the last couple days. Glad to see this has been resolved. Changes look great to me! Thanks!

**Pull request merged**